packages/components/src/isolated-event-container/README.md

```diff
@@ -0,0 +1,34 @@
+# Isolated Event Container
+
+This is a container that prevents certain events from propagating outside of the container. This is used to wrap
+UI elements such as modals and popovers where the propagated event can cause problems. The event continues to work
+inside the component.
+
+For example, a `mousedown` event in a modal container can propagate to the surrounding DOM, causing UI outside of the
+modal to be interacted with.
+
+The current isolated events are:
+- mousedown - This prevents UI interaction with other `mousedown` event handlers, such as selection
```

Comment:
How are we going to update this component later, will this be a breaking change on each new event handler? Can't we use a `events` prop instead to avoid making the decision in the component itself and leave it to the caller?

Reply:
I saw the component as internally containing all the events that might leak and cause problems for modal (or modal-ish) components. Passing them into the component moves that knowledge to each caller, and makes it more general purpose than planned.

I get your point about it potentially being a breaking change if a new event is added though. I don't see anything being added that would not be applicable to all users of the component, although that's easy to say without hindsight!

Maybe this is a naming thing. Would something like `<IsolatedMouseContainer />` narrow the scope to specifically handle mouse events? Any new non-mouse event would then require a seperate wrapper, along with separate tests.

Happy to change either way.

Reply:
Thanks for the clarification, it's fine for me. I don't have strong opinions on naming things. you call really.

Reply:
>I saw the component as internally containing all the events that might leak and cause problems for modal (or modal-ish) components.

At an abstract level, I like this idea. In practice, I don't see how it could possibly be implemented as described, at least not without (a) maintaining a full set of event handlers upon which to listen, and (b) providing handlers for _all_ those events (for events like `mousemove` and `scroll`, it becomes unfeasible).

I wonder if a reasonable compromise would be to put upon the consumer to provide the events for which propagation is stopped?

I could imagine it looking similar to `IgnoreNestedEvents`:

https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/ignore-nested-events/index.js

Reply:
Maybe we're not on the same page as to expectations of what IsolatedEventContainer is meant to address. Could you elaborate on its purpose?

>I saw the component as internally containing all the events that might leak and cause problems for modal (or modal-ish) components.

"all the events that might [...] cause problems" is a bit broad. What are those events? Couldn't _any_ event potentially cause some problem?

Reply:
That's very probable! The aim was to solve the specific instance of `mousedown` leaking out of modalized components in a more general way than adding the event handlers to each component.

Technically yes, other events could cause problems. Rather than cause unintended side-effects with a block-everything approach I thought it best to focus only on the known and tested event.

>  internally containing all the events that might leak

This was only meant in the sense that yes, any event _could_ cause a problem, somewhere and somehow, not that I was aware of them.

As I wondered above, maybe this could have been improved with more specific naming (and a clarified readme!) so it's clear that the component is only for this one case, and not a general purpose propagation blocker.

Also, after now being aware of `IgnoreNestedEvents` I will check the places it is used and see if the same events could cause problems with `IsolatedEventContainer`

Reply:
It was in my mind that rather than trying to plug holes as they come to surface (`mousedown` today, `keydown` tomorrow), if we could consider a modal to be a separate document / sandbox from the one in which its contained, it would be perfectly reasonable to stop _all_ events from escaping. This is where it comes to be a "catch-every-handler kind of thing", leading to the practical concerns outlined in my first comment.†

Ignoring that, and accepting the fact we have one-off handlers, it still seems like something which is unclear when contained within a component. In using `IsolatedEventContainer`, how am I as a developer to know what's being isolated? Having this explicit by a prop `events` could make this much clearer and provide an opportunity for the developer to _decide_ what it should handle, if they have some insight into anticipating the issues.

Conversely, there's a good chance they don't, and would benefit from some default set of handlers, closer to what's implemented now.

----

† As I write this, I think in my head "it's like an `iframe`, nothing escapes!" which then led to a crazy idea. What if it is an iframe? 😄 In fact, we have a good amount of the work done already with the [`Sandbox` component](https://github.com/WordPress/gutenberg/blob/master/packages/components/src/sandbox/README.md). I expect this might not be very feasible though, at least in trying to respect styling and allowing interactions (data operations) to occur between the iframe and the surrounding page.